utils/src/main/java/software/amazon/awssdk/utils/cache/lru/LruCache.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.cache.lru;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * An LRU (Least Recently Used) cache implementation.
+ * <p>
+ * The user can configure the maximum size of the cache, which is set to a default of 100.
+ */
+@SdkProtectedApi
+public final class LruCache<K, V>  {
+
+    private static final int DEFAULT_SIZE = 100;
+
+    private final Map<K, CacheEntry<K, V>> cache;
+    private final Function<K, V> valueSupplier;
+
+    private CacheEntry<K, V> leastRecentlyUsed = null;
+    private CacheEntry<K, V> mostRecentlyUsed = null;
+    private final Lock listLock = new ReentrantLock();
+    private final int maxCacheSize;
+
+    private LruCache(Builder<K, V> b) {
+        this.valueSupplier = b.supplier;
+        Integer customSize = Validate.isPositiveOrNull(b.maxSize, "size");
+        this.maxCacheSize = customSize != null ? customSize : DEFAULT_SIZE;
+        this.cache = new ConcurrentHashMap<>();
+    }
+
+    public V get(K key) {
+        CacheEntry<K, V> cachedEntry = cache.get(key);
+        if (cachedEntry != null) {
+            moveToFront(cachedEntry);
+            return cachedEntry.value();
+        }
+        V value = valueSupplier.apply(key);
+        CacheEntry<K, V> newEntry = new CacheEntry<>(key, value);
+        addToFront(newEntry);
+        return newEntry.value();
+    }
```

Comment:
Should we use `cache.computeIfAbsent` instead of `cache.get` and `cache.put`? That would lock around the key, so that the `valueSupplier` would only be executed once, even with concurrent `LruCache.get` calls.

Reply:
Yes, will refactor.